src/iop/highlights.c

```diff
@@ -1414,27 +1355,27 @@ static inline void heat_PDE_diffusion(const float *const restrict high_freq, con
 
         // fetch non-local pixels and store them locally and contiguously
         dt_aligned_pixel_t neighbour_pixel_HF[9];
-        dt_aligned_pixel_t neighbour_pixel_LF[9];
+        for_four_channels(c, aligned(neighbour_pixel_HF, HF: 64))
+        {
+          neighbour_pixel_HF[3 * 0 + 0][c] = HF[4 * (i_neighbours[0] + j_neighbours[0]) + c];
+          neighbour_pixel_HF[3 * 0 + 1][c] = HF[4 * (i_neighbours[0] + j_neighbours[1]) + c];
+          neighbour_pixel_HF[3 * 0 + 2][c] = HF[4 * (i_neighbours[0] + j_neighbours[2]) + c];
 
-        for(size_t ii = 0; ii < 3; ii++)
-          for(size_t jj = 0; jj < 3; jj++)
-          {
-            const size_t neighbor = 4 * (i_neighbours[ii] + j_neighbours[jj]);
-            const size_t nidx = 3 * ii + jj;
-            for_four_channels(c, aligned(neighbour_pixel_HF, HF, neighbour_pixel_LF, LF : 64))
-            {
-              neighbour_pixel_HF[nidx][c] = HF[neighbor + c];
-              neighbour_pixel_LF[nidx][c] = LF[neighbor + c];
-            }
-          }
+          neighbour_pixel_HF[3 * 1 + 0][c] = HF[4 * (i_neighbours[1] + j_neighbours[0]) + c];
+          neighbour_pixel_HF[3 * 1 + 1][c] = HF[4 * (i_neighbours[1] + j_neighbours[1]) + c];
+          neighbour_pixel_HF[3 * 1 + 2][c] = HF[4 * (i_neighbours[1] + j_neighbours[2]) + c];
+
+          neighbour_pixel_HF[3 * 2 + 0][c] = HF[4 * (i_neighbours[2] + j_neighbours[0]) + c];
+          neighbour_pixel_HF[3 * 2 + 1][c] = HF[4 * (i_neighbours[2] + j_neighbours[1]) + c];
+          neighbour_pixel_HF[3 * 2 + 2][c] = HF[4 * (i_neighbours[2] + j_neighbours[2]) + c];
+        }
 
         // Compute the laplacian in the direction parallel to the steepest gradient on the norm
-        float anisotropic_kernel_isophote[9];
-        compute_laplace_kernel(neighbour_pixel_LF, DIFFUSE_ISOPHOTE, anisotropic_kernel_isophote);
+        float anisotropic_kernel_isophote[9] = { 0.25f, 0.5f, 0.25f, 0.5f, -3.f, 0.5f, 0.25f, 0.5f, 0.25f };
 
         // Convolve the filter to get the laplacian
         dt_aligned_pixel_t laplacian_HF = { 0.f, 0.f, 0.f, 0.f };
-        for(size_t k = 0; k < 9; k++)
+        for(int k = 0; k < 9; k++)
         {
           for_each_channel(c, aligned(laplacian_HF, neighbour_pixel_HF, anisotropic_kernel_isophote: 64))
             laplacian_HF[c] += neighbour_pixel_HF[k][c] * anisotropic_kernel_isophote[k];
```

Comment:
laplacian_HF and neighbor_pixel_HF have 16-byte alignment, and anisotropic_kernel_isophote is currently not explicitly aligned at all (add DT_ALIGNED_ARRAY to its declaration).

Reply:
The fact that obviously wrong alignment declarations create no crash makes me wonder if these have any effect at all.

Reply:
Fixed.

Reply:
> The fact that obviously wrong alignment declarations create no crash makes me wonder if these have any effect at all.

Depends on the processor and which instructions the compiler generates (the older SSE instructions segfault on misalignment, the newer 256/512-bit instructions simply run slower).  For this particular loop, `anistropic_kernel_isophote` doesn't matter, since we're doing a scalar load and duplicating that across all channels.